File: beecrowd/2069 - A Mesa Quadrada de Inês Venezuela.cpp
#include <bits/stdc++.h>

using namespace std;

auto factors(int n) {
    vector<int> primes;

    for(int p = 2; p * p <= n; ++p) {
        if(n % p) continue;
        primes.push_back(p);
        while(n % p == 0) n /= p;
    }

    if(n > 1)
        primes.push_back(n);

    return primes;
}

int main() {
	ios_base :: sync_with_stdio(false);
	cin.tie(0);

    int a, b;

    cin >> a >> b;

    int c = gcd(a, b);
    long long L2 = 1;

    auto primes = factors(c);

    for(int p : primes) {
        int i = 0;

        while(c % p == 0) {
            c /= p;
            i ^= 1;
            L2 *= p;
        }

        if(i == 1) L2 *= p;
    }

    auto L = static_cast<long long>(sqrt(L2 + 0.0L));

    cout << L << '\n';

	return 0;
}
 

File: beecrowd/2069 - A Mesa Quadrada de Inês Venezuela.rs
#![allow(dead_code)]
use std::{
    io::{self, Read, Write},
    writeln,
};

struct Scanner {
    buffer: Vec<u8>,
    index: usize,
}

impl Scanner {
    fn new() -> Self {
        let mut input = Vec::new();

        io::stdin().read_to_end(&mut input).unwrap();

        Self {
            buffer: input,
            index: 0,
        }
    }

    fn has_next(&mut self) -> bool {
        while self.index < self.buffer.len() && self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        self.index < self.buffer.len()
    }

    fn next<T: std::str::FromStr>(&mut self) -> T {
        if !self.has_next() {
            panic!("End Of File");
        }

        let start = self.index;

        while self.index < self.buffer.len() && !self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        std::str::from_utf8(&self.buffer[start..self.index])
            .unwrap()
            .parse::<T>()
            .ok()
            .unwrap()
    }
}

fn gcd(mut a: i32, mut b: i32) -> i32 {
    while a != 0 {
        let temp = a;
        a = b % a;
        b = temp;
    }
    b
}

fn factors(mut n: i32) -> Vec<i32> {
    let mut primes = Vec::new();
    let mut p = 2;

    while p * p <= n {
        if n % p == 0 {
            primes.push(p);

            while n % p == 0 {
                n /= p;
            }
        }

        p += 1;
    }

    if n != 1 {
        primes.push(n);
    }

    primes
}

fn main() {
    let mut writer = io::BufWriter::with_capacity(1 << 20, io::stdout());

    let mut sc = Scanner::new();

    let a = sc.next::<i32>();
    let b = sc.next::<i32>();
    let mut c = gcd(a, b);
    let mut l2 = 1;

    let primes = factors(c);

    for p in primes {
        let mut odd = false;
        let prime = p as i64;

        while c % p == 0 {
            c /= p;
            odd = !odd;
            l2 *= prime;
        }

        if odd {
            l2 *= prime;
        }
    }

    let l = (l2 as f64).sqrt() as i64;

    writeln!(writer, "{}", l).ok();
}
